pkg/apis/core/types.go

```diff
@@ -2823,6 +2823,16 @@ type PodAffinityTerm struct {
 	// An empty selector ({}) matches all namespaces.
 	// +optional
 	NamespaceSelector *metav1.LabelSelector
+	// MatchLabelKeys is a set of pod label keys to select which pods will
+	// be taken into consideration. The keys are used to lookup values from the
+	// incoming pod labels, those key-value labels are ANDed with labelSelector
+	// to select the group of existing pods which pods will be taken into consideration
+	// for the incoming pod's pod (anti) affinity. Keys that don't exist in the incoming
+	// pod labels will be ignored. The default value is empty.
+	// This is an alpha field and requires enabling MatchLabelKeysInPodAffinity feature gate.
+	// +listType=atomic
+	// +optional
+	MatchLabelKeys []string
```

Comment:
Not about this PR, but relevant given that we are adding a new related field: document what does it mean when `LabelSelector` is null.

Reply:
@sanposhiho this can be good edge cases for UTs :)

Reply:
When the Label Selector in the Pod Affinity is empty, no objects will be matched.
That means if it's in a required PodAffinity, the Pod will never be schedulable.
and if it's in a required PodAntiAffinity, that PodAntiAffinity will be no-op.

Let me add this behavior on the doc after double-checking the behavior.

---

Also, I'm thinking it's very likely mistake and I've already created the issue to make the warning in such cases. 
https://github.com/kubernetes/kubernetes/issues/115873

Reply:
so if you just want to use MatchLabelKeys, users need to put an empty (but not nil) LabelSelector, right? I think that's consistent, but needs to be properly documented. Same for topology spreading.

Reply:
Unless we want to change the behavior here and in topology spreading so that if `MatchLabelKeys` is not empty, a nil LabelSelector is considered match-all? cc @denkensk

Reply:
> so if you just want to use MatchLabelKeys, users need to put an empty (but not nil) LabelSelector

Yes, right. So, the following PodAffinity can match any Pods.

```
  affinity:
    podAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        - labelSelector: {}
          topologyKey: kubernetes.io/hostname
```

and the following PodAffinity can never match any Pods.

```
  affinity:
    podAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        - topologyKey: kubernetes.io/hostname
```

---

> Unless we want to change the behavior here and in topology spreading so that if MatchLabelKeys is not empty, a nil LabelSelector is considered match-all

I'm thinking it's better to change.
From users PoV, it's easier to understand if the above two results in the same scheduling result: match-all.

Reply:
>  this can be good edge cases for UTs :)

Anyway, I'll add UTs for this behavior.

Reply:
@sanposhiho  Sorry. I think it's a bug in the feature MatchLabelKeys of TopologySpread. In case4 you listed, it needs to be no-match and ignored. I submit a pr to fix it in TopologySpread. https://github.com/kubernetes/kubernetes/pull/116535

Reply:
👍 
Yep, I agree that we should change the MatchLabelKeys implementation for consistency to just always make label value ANDed with LabelSelector.  I'll follow it in PodAffinity as well.

Then, the further question is "whether we want to change the way to handle null labelSelector" (as @alculquicondor said in https://github.com/kubernetes/kubernetes/pull/116065#discussion_r1132432361. )

Reply:
I'm feeling the opposite. 

To prevent mistakes, it may be useful to make the behaviors of `null` labelSelector and `{}` labelSelector the same (=match-all).

Let's think about the case a user wants to use just `matchLabelKeys` without any other `labelSelector`. 

After changing behavior in https://github.com/kubernetes/kubernetes/commit/11ebbb360ac8b8fa2c9903cd982890b3fcfd4c5e and https://github.com/kubernetes/kubernetes/pull/116535, he/she needs to write `labelSelector: {}` like:

```
  affinity:
    podAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        - labelSelector: {}
          topologyKey: kubernetes.io/hostname
          matchLabelKeys: 
          - "security"
```

But, if they mistakenly write like this (`null` labelSelector), then it's gonna be no-match.

```
  affinity:
    podAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        - topologyKey: kubernetes.io/hostname
          matchLabelKeys: 
          - "security"
```

So, I'm thinking that making the behaviors of `null` labelSelector and `{}` labelSelector the same (=match-all) looks like simpler and easier to users.

Reply:
memo: We'll forbid nil selector with non-nil matchLabelKeys in validation.
https://github.com/kubernetes/kubernetes/pull/116535#issuecomment-1467148341

Reply:
Add comment that if this is not empty, labelSelector must be set.